packages/chai-openapi-response-validator/test/assertions/satisfyApiSpec/openapi2SpecsDefineBasePath.test.js

```diff
@@ -0,0 +1,97 @@
+/** *****************************************************************************
+ * Copyright 2019 IBM Corp.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ****************************************************************************** */
+
+const chai = require('chai');
+const path = require('path');
+
+const chaiResponseValidator = require('../../..');
+
+const { expect, AssertionError } = chai;
+
+describe('Using OpenAPI 2 specs that define a basePath', () => {
+  before(() => {
+    chai.use(chaiResponseValidator(path.resolve(
+      '../../commonTestResources/exampleOpenApiFiles/valid/'
+         + 'openapi2DefineBasePath.yml'
+    )));
+  });
+
+  describe("res.req.path matches the basePath ('/test') and an endpoint path", () => {
+    const res = {
+      status: 200,
+      req: {
+        method: 'GET',
+        path: '/test/responseBody/string',
+      },
+      body: 'valid body (string)',
+    };
+
+    it('passes', () => {
+      expect(res).to.satisfyApiSpec;
+    });
+
+    it('fails when using .not', () => {
+      const assertion = () => expect(res).to.not.satisfyApiSpec;
+      expect(assertion).to.throw(AssertionError, '');
+    });
+  });
+  describe('res.req.path does not match any paths', () => {
+    const res = {
+      status: 200,
+      req: {
+        method: 'GET',
+        path: 'nonExistentServer/test/responseBody/string',
+      },
+      body: 'valid body (string)',
+    };
+
+    it('fails', () => {
+      const assertion = () => expect(res).to.satisfyApiSpec;
+      expect(assertion).to.throw(
+        "expected res to satisfy a '200' response defined for endpoint 'GET nonExistentServer/test/responseBody/string' in your API spec" +
+        "\nres had request path 'nonExistentServer/test/responseBody/string', but your API spec has no matching path" +
+                '\n\nPaths found in API spec: /test/responseBody/string'
```

Comment:
in our OpenAPI 3 serversDefined tests, you'll see that when we can't find a matching OpenAPI path, but we can find a `server` / `basePath`, we say `Servers found in API spec: ... `

As a user of OpenAPI 2, would you find it useful to see something like `basePath found (/test) but matches no basePath + serverURL combinations`?

That may be a bigger task, so I don't mind if we/I do that in a later PR

Reply:
Good question! I don't think the longer message adds much utility in this case. If you have multiple `servers`, you need to explicitly think about which one you're using. But with `basePath`, you're only adding it so you don't have to type it over and over again -- because you don't want to explicitly think about it. So I think it's fine to just say the URL as a whole is no good.

Reply:
Thanks for your reply, I agree.

Relatedly, do you think there's any risk that someone reading the error message
```
Paths found in API spec: /test/responseBody/string'
```
will be confused because they only have the `responseBody/string` path in the paths section of their spec?

Perhaps it's up to them to realise that we are combining their `basePath` with each path. Do you think we should provide a hint? (I think it's probably beyond the responsibility of this package).

Reply:
In fact, this decision affects OpenAPI 3, because the OpenAPI 2 `basePath` is similar to the OpenAPI 3 `server` URLs. I think the error message
```
Paths found in API spec: /test/responseBody/string'
```
should be consistent between OpenAPI 2 and 3, so let's decide now for both OpenAPI 2 and 3.

An OpenAPI 2 spec can have only one `basePath`. Suppose in the `Paths found in API spec` error message, we list paths without distinguishing `basePath` (`/test`) and the 'endpoint' paths (`responseBody/string`). A user who defines 10 endpoint paths in their OpenAPI 2 spec will see 10. User is happy.

But an OpenAPI 3 spec can have many server URLs. So suppose in `Paths found in API spec` error message, we list paths without distinguishing server URLs and 'endpoint' paths. A user who defines 10 endpoint paths in their OpenAPI 3 spec will see 10 `Paths found in API spec` if they have 1 server URL. But if they have 5 server URLs, they will see 50 `Paths found in API spec`.

So, questions for users @tgiardina and @JonnySpruce (if you're interested):
- for OpenAPI 3, should we list all 50 paths, or instead list 10 paths and then list 5 servers?
- must OpenAPI 2 be consistent with OpenAPI 3 on this?
- (what is least confusing for the user?)

Reply:
I'm not sure we can keep the error messages consistent, because I feel strongly about two points:

1) In your OpenApi 3 example, we should list 10 paths instead of 50 paths.
2) Users using OpenApi 2 should not have to decipher what `servers` are when reading an error message.

My guess is that everyone will agree on Point 1. I think the 50 paths option will obfuscate the error message quite a bit. Some internal code consistency is not worth the price.

Re Point 2, I think there is a lot of software that doesn't support OpenApi 3 yet (citation needed?) and so users who have only ever used OpenApi 2 should be treated as first class citizens. Part of being a first class citizen is getting errors messages specifically tailored for you to understand. And for example, someone who has never used OpenApi 3 may not be able to quickly and easily decipher:

```
'/basePath/nonExistentEndpointPath' matches servers [ '/basePath' ] but no <server/endpointPath> combinations
       
      -Servers found in API spec: /basePath`
```

If we can somehow revise the above error message so that it's consistent and adheres to both points above, let's do that. Otherwise, I think we should sacrifice a little bit of consistency for a whole lot of clarify.

Reply:
Okay, I agree with both points. Would these error messages satisfy them for OpenAPI 2 and 3, while keeping the two consistent?

OpenAPI 2:

When request path does not match your `basePath`:
```
expected res to satisfy a '200' response defined for endpoint 'GET /wrongBasePath' in your API spec

res had request path '/wrongBasePath', but your API spec has basePath '/basePath'
```

When request path matches your `basePath` but no endpoint paths:
```
expected res to satisfy a '200' response defined for endpoint 'GET basePath/nonExistentEndpointPath' in your API spec

res had request path '/basePath/nonExistentEndpointPath' but your API spec has no matching path

Paths found in API spec: /endpointPath

'/basePath/nonExistentEndpointPath' matches basePath `/basePath` but no <basePath/endpointPath> combinations
```


OpenAPI 3:

When request path does not match any of your `server`s (this removes the `Paths found in API spec` lines from the existing message):
```
expected res to satisfy a '200' response defined for endpoint 'GET /nonExistentServer' in your API spec

res had request path '/nonExistentServer', but your API spec has no matching servers

Servers found in API spec: /server1, /server2, /server3
```

When request path matches some of your `server`s but no endpoint paths:
```
expected res to satisfy a '200' response defined for endpoint 'GET /server1/nonExistentEndpointPath' in your API spec

res had request path '/server1/nonExistentEndpointPath' but your API spec has no matching path

Paths found in API spec: /test

'/server1/nonExistentEndpointPath' matches servers '/server1' but no <server/endpointPath> combinations

Servers found in API spec: /server1, /server2, /server3
```

Reply:
@rwalle61, I think the proposal above looks good to me, the only thing I would add potentially change is the OpenAPI 3 no endpoint paths message. I would simply remove the additional information about servers found, since you've already matched one and so that is probably information you don't need, and also keeps the structure consistent with OpenAPI 2:

```
expected res to satisfy a '200' response defined for endpoint 'GET /server1/nonExistentEndpointPath' in your API spec

res had request path '/server1/nonExistentEndpointPath' but your API spec has no matching path

Paths found in API spec: /test

'/server1/nonExistentEndpointPath' matches servers '/server1' but no <server/endpointPath> combinations
```

Reply:
@rwalle61, The proposed solution messages look good to me! I also agree with everything @JonnySpruce said! Although I will add that I don't think the additional `servers` information will confuse anyone. It's just sorta weird and unnecessary.

Let me know when you've solidified your choice, and I will wrap up this PR!

Reply:
Thanks @JonnySpruce and @tgiardina, let's go with those messages (removing the `servers` info for OpneAPI 3)!